crypto/err/err.c

```diff
@@ -895,3 +896,19 @@ void err_clear_last_constant_time(int clear)
                                      0, ERR_FLAG_CLEAR);
     es->err_flags[top] |= clear;
 }
+
+void *ERR_raise_malloc_failure(void *ptr,
+                               const char *file, int line, const char *func)
+{
+    if (ptr != NULL)
+        return ptr; /* all good */
+
+    /*
+     * Mem alloc error loop while reporting malloc error is prevented by using
+     * CRYPTO_zalloc() rather than OPENSSL_zalloc for ERR_STATE allocation-
+     */
+    ERR_new();
+    ERR_set_debug(file, line, func);
+    ERR_set_error(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE, NULL);
```

Comment:
Shouldn't the caller say what lib it is, rather than having this hard-coded `ERR_LIB_CRYPTO`?  In other words, you will need another argument `int lib` to this function.

Reply:
Unfortunately that argument is nowhere to get from. The caller cannot supply it as OPENSSL_malloc() and friends do not have such argument and adding it would mean adding a new set of macros with this argument which would be really ugly.

So there is still some small regression of the error data reported but IMO this is still acceptable.

Reply:
So basically, any library that may want to show where the memory allocation failure happened at a higher level will have to trade their current:

``` C
    ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE);
```

For something like this:

``` C
    ERR_raise(ERR_LIB_FOO, ERR_R_CRYPTO_LIB);
```

'cause you see, some allocation failures may be caused because the ask is too bit (to much memory demand), and it may be valuable to know where that ask was made!

Reply:
Why? So far, this PR just adds extra calls that are equivalent to `ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE)`.
All existing `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` can stay as they are.
The idea was to remove them in some subsequent PRs, but we can also keep (part of) them.

Reply:
> Why?

Not sure which part you're asking about...

> All existing `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` can stay as they are. The idea was to remove them in some subsequent PRs, but we can also keep (part of) them.

So right now, with this PR, we'll get double reports any time there's a memory allocation failure, in the case the caller reports it.  Of course, that will be seen as unnecessary, so a natural follow-up will most probably be to remove the reports in the caller (and losing some info of where it happened), or when someone does want to leave a trace in the caller, will probably change it as I suggested above (the ERR_R_FOO_LIB reasons have the same reason number as the ERR_LIB_FOO subsystem numbers, and are a way of saying that "something went wrong in subsystem FOO", along with info about the caller).

In the end, this gets back to us who develop OpenSSL.  The subsystem name, function name, file name and line number are all bits of a puzzle to find where something went wrong.  At the same time, because code moves around, each piece of information can be considered fuzzy, making it increasingly difficult to figure out causes for failures, especially if the error report isn't your own, but from a user raising an issue.  In my mind, every bit of information counts, which is the main reason why I'm having these discussions (honestly, feels like fighting with teeth and nails at times)

Reply:
> > Why?
> 
> Not sure which part you're asking about...

I was questioning this:
>> So basically, any library that may want to show where the memory allocation failure happened at a higher level will have to trade their current:
>> `    ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE);`
>> For something like this:
>>`     ERR_raise(ERR_LIB_FOO, ERR_R_CRYPTO_LIB);`

because I did not see (but see below) much need for such changes - as I continued to write:

> > All existing `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` can stay as they are.


>> The idea was to remove them in some subsequent PRs, but we can also keep (part of) them.
> 
> So right now, with this PR, we'll get double reports any time there's a memory allocation failure, in the case the caller reports it.

Well, not strictly duplicates. The caller uses 
```
ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)
```
where `FOO` usually is not `CRYPTO`,
while the new reporting in `mem.c` essentially does
```
ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE)
```
and in any case the file, line, and func info are different.

> Of course, that will be seen as unnecessary, so a natural follow-up will most probably be to remove the reports in the caller (and losing some info of where it happened),

Yep, this was the immediate reflex, but as I meanwhile wrote elsewhere here, I appreciate the idea of leaving traces - where it is worth doing so. So in the end it is a choice to be take for each existing `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` whether to retain it or not.

> or when someone does want to leave a trace in the caller, will probably change it as I suggested above (the ERR_R_FOO_LIB reasons have the same reason number as the ERR_LIB_FOO subsystem numbers, and are a way of saying that "something went wrong in subsystem FOO", along with info about the caller).

Which this explanation of yours, I now understand that the `ERR_raise(ERR_LIB_FOO, ERR_R_CRYPTO_LIB)` is a nice way of adding a trace point without committing to what (supposedly) went wrong at lower level in CRYPTO.
 
> In the end, this gets back to us who develop OpenSSL. The subsystem name, function name, file name and line number are all bits of a puzzle to find where something went wrong. At the same time, because code moves around, each piece of information can be considered fuzzy, making it increasingly difficult to figure out causes for failures when there's less of said information, especially if the error report isn't your own, but from a user raising an issue. In my mind, every bit of information counts, which is the main reason why I'm having these discussions (honestly, feels like fighting with teeth and nails at times)

Yeah, as I wrote yesterday, it would be nice if C offered a convenient way of automatically providing a stack/caller trace, like for instance Java does.

For the specific case of `ERR_R_MALLOC_FAILURE` I doubt that it is worth manually adding or retaining such more higher-level trace points, simply because such failures are rather rare. I believe it is very rare that an malloc caller by mistake asked for an unrealistic amount of memory. And if the request was reasonable but could not be fulfilled because of actual memory shortage in the run time system, the higher-level trace points are pretty irrelevant. 
To make things worse, so far there are many `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` invocations where the caller just got a NULL pointer from a function call just done and presumes that the reason was an malloc failure, while in fact something else went wrong (such as some missing initialization) and the `ERR_R_MALLOC_FAILURE` is even misleading.

To sum up, while I do see the merits of higher-level trace points for other types or errors/failures, I do not see it so much for malloc failures. Yet in order to avoid having to decide for each and every existing `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` whether to keep it or not, I meanwhile like the idea of simply replacing all of them by `ERR_raise(ERR_LIB_FOO, ERR_R_CRYPTO_LIB)`.

Reply:
I am not overly concerned about the library the error comes from: these are malloc failures: bad allocation size, out of memory, overly fragmented memory or memory corruption.

We've got file, line and maybe function.  The file is often enough to identify the library.